cpp/ql/src/semmle/code/cpp/valuenumbering/HashCons.qll

```diff
@@ -0,0 +1,519 @@
+/**
+ * Provides an implementation of Hash consing.
+ * See https://en.wikipedia.org/wiki/Hash_consing
+ *
+ * The predicate `hashCons` converts an expression into a `HC`, which is an
+ * abstract type presenting the hash-cons of the expression. If two
+ * expressions have the same `HC` then they are structurally equal.
+ *
+ * Important note: this library ignores the possibility that the value of
+ * an expression might change between one occurrence and the next. For
+ * example:
+ *
+ * ```
+ * x = a+b;
+ * a++;
+ * y = a+b;
+ * ```
+ *
+ * In this example, both copies of the expression `a+b` will hash-cons to
+ * the same value, even though the value of `a` has changed. This is the
+ * intended behavior of this library. If you care about the value of the
+ * expression being the same, then you should use the GlobalValueNumbering
+ * library instead.
+ *
+ * To determine if the expression `x` is structurally equal to the
+ * expression `y`, use the library like this:
+ *
+ * ```
+ * hashCons(x) = hashCons(y)
+ * ```
+ */
+
+/*
+ * Note to developers: the correctness of this module depends on the
+ * definitions of HC, hashCons, and analyzableExpr being kept in sync with
+ * each other. If you change this module then make sure that the change is
+ * symmetric across all three.
+ */
+
+import cpp
+
+/** Used to represent the hash-cons of an expression. */
+private cached newtype HCBase =
+  HC_IntLiteral(int val, Type t) { mk_IntLiteral(val,t,_) }
+  or
+  HC_FloatLiteral(float val, Type t) { mk_FloatLiteral(val,t,_) }
+  or
+  HC_StringLiteral(string val, Type t) {mk_StringLiteral(val,t,_)}
+  or
+  HC_Variable(Variable x) {
+     mk_Variable(x, _)
+  }
+  or
+  HC_FieldAccess(HC s, Field f) {
+    mk_DotFieldAccess(s,f,_) or
+    mk_PointerFieldAccess_with_deref(s,f,_) or
+    mk_ImplicitThisFieldAccess_with_deref(s,f,_)
+  }
+  or
+  HC_Deref(HC p) {
+    mk_Deref(p,_) or
+    mk_PointerFieldAccess(p,_,_) or
+    mk_ImplicitThisFieldAccess_with_qualifier(p,_,_)
+  }
+  or
+  HC_ThisExpr(Function fcn) {
+    mk_ThisExpr(fcn,_) or
+    mk_ImplicitThisFieldAccess(fcn,_,_)
```

Comment:
I think that's a holdover from the GVN library. @kevinbackhouse can you confirm?

Reply:
Yes, that's correct.

Reply:
It looks like implicit `this` is usually expanded in the extractor, so it may be better to leave that in place and add some notes about it, rather than making the treatment of an implicit `this` be inconsistent.

Reply:
It's rare to see an implicit `this` in the AST. I think it happens only for calls to the destructors of fields in compiler-generated destructors. I don't think that's something we need to worry about for HashCons.

Reply:
I still think it's worth simplifying the hash-consing of field accesses. I don't think it's important that `myField` gets equated with `this.myField` as those two would never occur in the same function because the extractor inserts explicit `this.` in user-written code.

Reply:
Yeah, if it's that rare I'll take the logic out;.